### ISSUE TITLE
Convert trees to a specific format

### ISSUE THREAD

**Issue by softwaredeng** [dmlc/xgboost#2156]
Trying to enable the inTrees package for the xgboost R package, for rule extraction and analysis. 

My question is that, how to convert an xgboost tree to the following format (which randomForest uses)? thanks!

> library(randomForest)
> rf <- randomForest(Species ~ ., data=iris)
> getTree(rf, 1, labelVar=TRUE)
node   left daughter right daughter    split var split point status prediction
1              2              3 Petal.Length        4.75      1       <NA>
2              4              5 Sepal.Length        5.45      1       <NA>
3              6              7  Sepal.Width        3.15      1       <NA>
4              8              9  Petal.Width        0.80      1       <NA>
5             10             11  Sepal.Width        3.60      1       <NA>
6              0              0         <NA>        0.00     -1  virginica
7             12             13  Petal.Width        1.90      1       <NA>
8              0              0         <NA>        0.00     -1     setosa
9             14             15  Petal.Width        1.55      1       <NA>
10             0              0         <NA>        0.00     -1 versicolor
11             0              0         <NA>        0.00     -1     setosa
12            16             17 Petal.Length        5.40      1       <NA>
13             0              0         <NA>        0.00     -1  virginica
14             0              0         <NA>        0.00     -1 versicolor
15             0              0         <NA>        0.00     -1  virginica
16             0              0         <NA>        0.00     -1 versicolor
17             0              0         <NA>        0.00     -1  virginica

**Comment by GillesVandewiele** [comment]
You can execute xgb_model.clf._Booster.get_dump() to get a string representation of all trees in the ensemble. Unfortunately, xgb inherently works with leaf values instead of class probabilities at each leaf, thus you will need to convert those leaf values to probabilities in some hacky way (https://github.com/dmlc/xgboost/issues/1746#issuecomment-290130695)

I once wrote code that converts this output from R randomForest to my own decision tree object (https://github.com/IBCNServices/GENESIM/blob/master/constructors/inTrees.py). Lately, I did the same for the xgb trees.

`
from sklearn.cross_validation import StratifiedKFold
from xgboost import plot_tree
import matplotlib.pyplot as plt

from constructors.ensemble import XGBClassification
from data.load_all_datasets import load_all_datasets
from sortedcontainers import SortedList, SortedSet

import numpy as np
from collections import Counter
import time

from decisiontree import DecisionTree


def parse_xgb_tree_string(tree_string, training_data, feature_cols, label_col, the_class):
    # Get class distribution
    _classes = np.unique(training_data[label_col].values)
    class_distributions = {}
    for _class in _classes:
        data = training_data[training_data[label_col] != _class]
        class_counts = Counter(data[label_col].values)
        total = sum(class_counts.values(), 0.0)
        for key in class_counts:
            class_counts[key] /= total
        class_distributions[_class] = class_counts

    # Get the unique values per feature
    unique_values_per_feature = {}
    for feature_col in feature_cols:
        # Just use a simple sorted list of np.unique (faster than SortedList or a set to get index of elements after testing)
        unique_values_per_feature[feature_col] = sorted(np.unique(training_data[feature_col].values))

    return parse_xgb_tree(tree_string, _class=the_class, class_distributions=class_distributions,
                          unique_values_per_feature=unique_values_per_feature, n_samples=len(training_data))


def get_closest_value(x, values):
    for i in range(len(values)-1):
        if values[i+1] > x:
            return float(values[i])
    return x


def parse_xgb_tree(tree_string, _class=0, class_distributions={}, unique_values_per_feature={}, n_samples=0):
    # There is some magic involved! The leaf values need to be converted to class distributions somehow!
    # For binary classification problems: convert to probability by calculating 1/(1+exp(-value))
    # For multi_class: the tree_string contains n_estimators * n_classes decision trees
    # WARNING: Classes are sorted according to output of np.unique
    # Ordered as follows [tree_1-class_1, ..., tree_1-class_k, tree_2-class_1, ....]

    # The problem is: tree_i is different for each class...
    # One possibility is to assign the probability to that class by calculating logistic function
    # And dividing the rest of the probability (sum to 1) according to the distribution of the remaining classes

    # Next problem is: everything is expressed as "feature < threshold" instead "feature <= threshold"
    # Solution is: take the infimum of those feature values

    decision_trees = {}
    # Binary classification
    binary_classification = len(class_distributions.keys()) == 2
    for line in tree_string.split('\n'):
        if line != '':
            _id, rest = line.split(':')
            _id = _id.lstrip()
            if rest[:4] != 'leaf':
                feature = rest.split('<')[0][1:]
                highest_lower_threshold = get_closest_value(float(rest.split('<')[1].split(']')[0]),
                                                            unique_values_per_feature[feature])
                decision_trees[_id] = DecisionTree(right=None, left=None,
                                                   label=feature, value=highest_lower_threshold,
                                                   parent=None)
            else:
                leaf_value = float(rest.split('=')[1])
                if binary_classification:
                    probability = 1/(1+np.exp(leaf_value))
                    other_class = class_distributions[_class].keys()[0]
                    class_probs = {_class: int(n_samples*probability),
                                   other_class: int(n_samples*(1-probability))}
                    if probability > 0.5: most_probable_class = _class
                    else: most_probable_class = other_class
                else:
                    probability = 1/(1+np.exp(-leaf_value))
                    class_probs = {}
                    remainder_samples = int(n_samples - probability*n_samples)
                    class_probs[_class] = int(probability*n_samples)
                    most_probable_class, most_samples = _class, class_probs[_class]
                    for other_class in class_distributions[_class]:
                        amount_samples = int(remainder_samples*class_distributions[_class][other_class])
                        class_probs[other_class] = amount_samples
                        if amount_samples > most_samples:
                            most_probable_class = other_class
                            most_samples = amount_samples

                decision_trees[_id] = DecisionTree(right=None, left=None,
                                                   label=most_probable_class, value=None,
                                                   parent=None)
                decision_trees[_id].class_probabilities = class_probs

    # Make another pass to link the different decision trees together
    for line in tree_string.split('\n'):
        if line != '':
            _id, rest = line.split(':')
            _id = _id.lstrip()
            tree = decision_trees[_id]
            if rest[:4] != 'leaf':
                rest = rest.split(']')[1].lstrip()
                links = rest.split(',')
                for link in links:
                    word, link_id = link.split('=')
                    if word == 'yes' or word == 'missing':
                        tree.left = decision_trees[link_id]
                    else:
                        tree.right = decision_trees[link_id]

    return decision_trees['0']

NR_FOLDS = 5
xgb = XGBClassification()
for dataset in load_all_datasets():
    df = dataset['dataframe']
    label_col = dataset['label_col']
    feature_cols = dataset['feature_cols']

    skf = StratifiedKFold(df[label_col], n_folds=NR_FOLDS, shuffle=True, random_state=1337)

    for fold, (train_idx, test_idx) in enumerate(skf):
        print 'Fold', fold + 1, '/', NR_FOLDS, 'for dataset', dataset['name']
        train = df.iloc[train_idx, :].reset_index(drop=True)
        X_train = train.drop(label_col, axis=1)
        y_train = train[label_col]
        test = df.iloc[test_idx, :].reset_index(drop=True)
        X_test = test.drop(label_col, axis=1)
        y_test = test[label_col]

        xgb_model = xgb.construct_classifier(train, feature_cols, label_col)

        n_classes = len(np.unique(y_train.values))
        if n_classes > 2:
            for idx, tree_string in enumerate(xgb_model.clf._Booster.get_dump()):
                tree = parse_xgb_tree_string(tree_string, train, feature_cols, label_col, np.unique(y_train.values)[idx % n_classes])
                tree.visualise('xgbtree')
                plot_tree(xgb_model.clf, num_trees=idx)
                plt.show()
                raw_input()
        else:
            for idx, tree_string in enumerate(xgb_model.clf._Booster.get_dump()):
                tree = parse_xgb_tree_string(tree_string, train, feature_cols, label_col, 0)
                tree.visualise('xgbtree')
                plot_tree(xgb_model.clf, num_trees=idx)
                plt.show()
                raw_input()
`

**Comment by softwaredeng** [comment]
thanks @GillesVandewiele! Will take a further look.

**Comment by GillesVandewiele** [comment]
More than welcome! Love your work on inTrees btw! Made a python wrapper around it :)

**Comment by khotilov** [comment]
@softwaredeng here's an example (after peeking at your GBM2List and formatGBM functions):
```R
library(xgboost)
library(data.table)

dtrain <- xgb.DMatrix(as.matrix(iris[, -5]), label = as.numeric(iris[, "Species"]) - 1)
watchlist <- list( train = dtrain)
param <- list(max_depth = 5, eta = 0.2, nthread = 2, objective = "multi:softmax", 
    num_class = 3, eval_metric = "mlogloss")
set.seed(1)
bst <- xgb.train(param, dtrain, nrounds = 10, watchlist)

xt <- xgb.model.dt.tree(feature_names = as.character(1:ncol(dtrain)), model=bst)
xt[Feature == 'Leaf', Feature := '-1']
xt[, 'split var' := as.integer(Feature)]
xt[, 'split point' := ifelse(`split var` > -1, Split, Quality)]
xt[, 'left daughter' := as.integer(tstrsplit(Yes, '-')[[2]]) + 1]
xt[, 'right daughter' := as.integer(tstrsplit(No, '-')[[2]]) + 1]
xt[, MissingNode := as.integer(tstrsplit(Missing, '-')[[2]]) + 1]
xt[, Weight := Cover]
xt[, Prediction := Quality]
xt[, Node := Node + 1]
xt[, c('ID', 'Yes', 'No', 'Split','Missing', 'Quality', 'Cover', 'Feature') := NULL]
for (f in c('left daughter', 'right daughter', 'MissingNode'))
  set(xt, which(is.na(xt[[f]])), f, -1)
xt[Tree<3]
```
```
    Tree Node split var   split point left daughter right daughter MissingNode    Weight    Prediction
 1:    0    1         3  2.450000e+00             2              3           2 66.666664  7.229677e+01
 2:    0    2        -1  2.870813e-01            -1             -1          -1 22.222221  2.870813e-01
 3:    0    3        -1 -1.466993e-01            -1             -1          -1 44.444443 -1.466993e-01
 4:    1    1         3  2.450000e+00             2              3           2 66.666664  1.807419e+01
 5:    1    2        -1 -1.435407e-01            -1             -1          -1 22.222221 -1.435407e-01
 6:    1    3         4  1.750000e+00             4              5           4 44.444443  4.190784e+01
 7:    1    4         3  4.950000e+00             6              7           6 23.999998  4.589849e+00
 8:    1    5         3  4.850000e+00             8              9           8 20.444443  6.351590e-01
 9:    1    6         4  1.550000e+00            10             11          10 21.333332  2.558654e-01
10:    1    7         4  1.550000e+00            12             13          12  2.666667  8.571428e-01
11:    1    8        -1 -5.108970e-09            -1             -1          -1  1.333333 -5.108970e-09
12:    1    9        -1 -1.425415e-01            -1             -1          -1 19.111111 -1.425415e-01
13:    1   10        -1  2.857143e-01            -1             -1          -1 19.999998  2.857143e-01
14:    1   11        -1  8.571428e-02            -1             -1          -1  1.333333  8.571428e-02
15:    1   12        -1 -8.571430e-02            -1             -1          -1  1.333333 -8.571430e-02
16:    1   13        -1  8.571428e-02            -1             -1          -1  1.333333  8.571428e-02
17:    2    1         4  1.650000e+00             2              3           2 66.666664  5.972297e+01
18:    2    2         3  4.950000e+00             4              5           4 45.333332  5.965446e+00
19:    2    3        -1  2.686567e-01            -1             -1          -1 21.333332  2.686567e-01
20:    2    4        -1 -1.465995e-01            -1             -1          -1 43.111107 -1.465995e-01
21:    2    5        -1  1.448276e-01            -1             -1          -1  2.222222  1.448276e-01
```
I assume you can take it from here and define the `status` column as you see appropriate. Keep in mind that unlike gbm, xgboost passes missing values either to the right or to the left (whichever is more appropriate according to data).

**Comment by softwaredeng** [comment]
@GillesVandewiele That's great! I'll link to your project.

**Comment by softwaredeng** [comment]
@khotilov nice, thank you for providing an elegant way producing the matrix!

**Comment by Zephyr1213** [comment]
@softwaredeng  I'm trying to hook up your InTrees with xgboost too, have you made any progress on that? or are you available to answer some questions I'm running into?

**Comment by GillesVandewiele** [comment]
@Zephyr1213 where are you having troubles with? Extracting decision trees from XGBoost or feeding the trees to inTrees? Maybe I can help...

**Comment by softwaredeng** [comment]
@Zephyr1213 @GillesVandewiele I was distracted by other stuff but it is on my agenda. Yes, let me know if there are any questions.

**Comment by Zephyr1213** [comment]
@GillesVandewiele  I used the codes @khotilov posted above to extract the trees from xgboost, it seems I need to edit @softwaredeng 's RF2List or GBM2List function to make everything click, right? I was trying to do that, but haven't got time to test it out yet.

**Comment by softwaredeng** [comment]
@Zephyr1213  correct - create a function something like XGB2List. I have added you to the inTrees github project. feel free to push your code there, and I can take a look. 

@GillesVandewiele added the python wrapper link to the project.

**Comment by qingyuanxingsi** [comment]
@GillesVandewiele Do your python wrapper support xgboost now??

**Comment by GillesVandewiele** [comment]
@qingyuanxingsi it does :). Make sure to take a look at https://github.com/IBCNServices/GENESIM/blob/master/constructors/genesim.py#L427 etc (it's been quite a while that I actively used/developed GENESIM)

Make an issue if you have any troubles or whatsoever :).

**Comment by softwaredeng** [comment]
close this issue. xgboost is now added to inTrees, should be on cran soon. https://github.com/softwaredeng/inTrees/blob/master/inTrees_1.2.tar.gz
